Lib/functools.py

```diff
@@ -273,57 +273,151 @@ def reduce(function, sequence, initial=_initial_missing):
+    pto_args = self.args
+    if pto_phcount:
+        n = len(args)
+        if n < pto_phcount:
+            raise TypeError(
+                "missing positional arguments "
+                "in 'partial' call; expected "
+                f"at least {pto_phcount}, got {n}")
+        pto_args = list(pto_args)
+        pos = j = 0
+        while j < pto_phcount:
+            pos = pto_args.index(Placeholder, pos)
+            pto_args[pos] = args[j]
+            pos += 1
+            j += 1
+        args = args[pto_phcount:] if n > pto_phcount else ()
+    keywords = {**self.keywords, **keywords}
+    return pto_args, args, keywords
+
+
+def _partial_repr(self):
+    cls = type(self)
+    module = cls.__module__
+    qualname = cls.__qualname__
+    args = [repr(self.func)]
+    args.extend(map(repr, self.args))
+    args.extend(f"{k}={v!r}" for k, v in self.keywords.items())
+    return f"{module}.{qualname}({', '.join(args)})"
+
+
 # Purely functional, no descriptor behaviour
 class partial:
     """New function with partial application of the given arguments
     and keywords.
     """
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    __slots__ = ("func", "args", "keywords", "placeholder_count",
+                 "__dict__", "__weakref__")
 
     def __new__(cls, func, /, *args, **keywords):
         if not callable(func):
             raise TypeError("the first argument must be callable")
-
-        if isinstance(func, partial):
-            args = func.args + args
-            keywords = {**func.keywords, **keywords}
-            func = func.func
-
-        self = super(partial, cls).__new__(cls)
-
+        func, args, kwds, phcount = _partial_prepare_new(cls, func, args,
+                                                         keywords)
+        self = super().__new__(cls)
         self.func = func
         self.args = args
-        self.keywords = keywords
+        self.keywords = kwds
+        self.placeholder_count = phcount
         return self
 
     def __call__(self, /, *args, **keywords):
-        keywords = {**self.keywords, **keywords}
-        return self.func(*self.args, *args, **keywords)
+        pargs, args, kwds = _partial_prepare_call(self, args, keywords)
+        return self.func(*pargs, *args, **kwds)
 
-    @recursive_repr()
-    def __repr__(self):
-        cls = type(self)
-        qualname = cls.__qualname__
-        module = cls.__module__
-        args = [repr(self.func)]
-        args.extend(repr(x) for x in self.args)
-        args.extend(f"{k}={v!r}" for (k, v) in self.keywords.items())
-        return f"{module}.{qualname}({', '.join(args)})"
+    __repr__ = recursive_repr()(_partial_repr)
 
     def __reduce__(self):
-        return type(self), (self.func,), (self.func, self.args,
-               self.keywords or None, self.__dict__ or None)
+        state = (
+            self.func,
+            self.args,
+            self.keywords or None,
+            self.placeholder_count,
```

Comment:
This is not compatible with older versions, even if there are no placeholders.

Is it necessary to save `placeholder_count`? Cannot it be deduced from `args`, as in the constructor?